quoridor-logic-server/src/main/java/com/rl/mpquoridor/models/board/GameBoard.java

```diff
@@ -0,0 +1,355 @@
+        this.getPhysicalBoard().putWall(wall); // Putting this, if pawn has no path, remove this wall.
+        if(!isAllPawnsHavePath()) {
+            this.getPhysicalBoard().removeWall(wall); // removing the illegal wall.
+            throw new IllegalMovementException("Not all pawns have available path to their end");
+        }
+
+        this.getPhysicalBoard().reduceWallToPawn(pawn);
+    }
+
+    public void executeAction(Pawn pawn, TurnAction action) throws  IllegalMovementException{
+        if(action instanceof PlaceWall) {
+            placeWall(pawn, ((PlaceWall) action).getWall());
+        } else if(action instanceof MovePawn) {
+            movePawn(pawn, ((MovePawn) action).getDirection());
+        }
+    }
+
+    /**
+     *
+     * @return a readonly copy of the physical board
+     */
+    public PhysicalBoard getPhysicalBoard() {
+        if(this.readOnlyPhysicalBoard == null) {
+            this.readOnlyPhysicalBoard = new ReadOnlyPhysicalBoard(this.board);
+        }
+        return this.readOnlyPhysicalBoard;
+    }
+
+    /**
+     * Should be an O(1) implementation.
+     *  @return the winner pawn or null
+     */
+    public Pawn getWinner() {
+        return this.winner;
+    }
+
+    private boolean isPathExists(Position s, HashSet<Position> dest) {
+        Queue<Position> q = new LinkedList<>();
+        Set<Position> visited = new HashSet<>();
+        q.add(s);
+        while(!q.isEmpty()) {
+            Position curr = q.poll();
+            if(dest.contains(curr)) { // This is an O(1) operation
+                return true;
+            }
+            if(!visited.contains(curr)) {
+                visited.add(curr);
+                q.addAll(getAvailableNeighbors(curr));
+            }
+        }
+
+        return false;
+    }
+
+    // todo: Change this to work with other dataSet
+    private List<Position> getAvailableNeighbors(Position p) {
+        List<Position> ret = new ArrayList<>();
+        // UP
+        if(p.getI() > 0) {
+            if (!this.getPhysicalBoard().getWalls().contains(new Wall(new Position(p.getI(), p.getJ()), WallDirection.RIGHT)) &&
+                    !this.getPhysicalBoard().getWalls().contains(new Wall(new Position(p.getI(), p.getJ() - 1), WallDirection.RIGHT))){
+                ret.add(new Position(p.getI() - 1, p.getJ()));
+            }
+        }
+        // DOWN
+        if(p.getI() < this.getPhysicalBoard().getSize() - 1) {
+            if (!this.getPhysicalBoard().getWalls().contains(new Wall(new Position(p.getI()+1, p.getJ()), WallDirection.RIGHT)) &&
+                    !this.getPhysicalBoard().getWalls().contains(new Wall(new Position(p.getI()+1, p.getJ() - 1), WallDirection.RIGHT))){
+                ret.add(new Position(p.getI() + 1, p.getJ()));
+            }
+        }
+
+        // LEFT
+        if(p.getJ() > 0) {
+            if (!this.getPhysicalBoard().getWalls().contains(new Wall(new Position(p.getI(), p.getJ()), WallDirection.DOWN)) &&
```

Comment:
If you convert all the calling of this method to **checkIfWallExist(i,j,direction)** it will make the class more readable